payjoin-cli/src/app/v2.rs

```diff
@@ -335,26 +375,148 @@ fn try_contributing_inputs(
+            payjoin::io::fetch_ohttp_keys(selected_relay.clone(), payjoin_directory.clone()).await
         };
-        #[cfg(not(feature = "_danger-local-https"))]
-        let ohttp_keys = payjoin::io::fetch_ohttp_keys(ohttp_relay, payjoin_directory).await?;
-        Ok(ohttp_keys)
+
+        match ohttp_keys {
+            Ok(keys) => return Ok(Some(keys)),
+            Err(error) => {
+                let status_code = error
+                    .to_string()
+                    .split_whitespace()
+                    .find_map(|code| http::StatusCode::from_str(code).ok())
+                    .expect("Could not find status code");
+                if !&status_code.is_success() {
+                    log::debug!("{error:?}");
+                    return Err(error.into());
+                } else {
+                    log::debug!("Failed to connect to relay: {selected_relay}, {error:?}");
+                    relay_state.lock().unwrap().add_failed_relay(selected_relay);
+                }
+            }
+        }
+    }
+}
+
+//Local relays are incapable of acting as proxies so we must opportunistically fetch keys from the config
+#[cfg(feature = "_danger-local-https")]
+async fn fetch_keys(
+    config: &Config,
+    _relay_state: Arc<Mutex<RelayState>>,
+) -> Result<Option<payjoin::OhttpKeys>> {
+    let keys = config.v2()?.ohttp_keys.clone().expect("No OHTTP keys set");
+
+    Ok(Some(keys))
+}
+
+#[cfg(not(feature = "_danger-local-https"))]
+async fn validate_relay(
+    config: &Config,
+    relay_state: Arc<Mutex<RelayState>>,
+) -> Result<payjoin::Url> {
+    use std::str::FromStr;
+
+    use payjoin::bitcoin::secp256k1::rand::prelude::SliceRandom;
+    let payjoin_directory = config.v2()?.pj_directory.clone();
+    let relays = config.v2()?.ohttp_relays.clone();
+
+    loop {
+        let failed_relays = relay_state.lock().unwrap().get_failed_relays();
+
+        let remaining_relays: Vec<_> =
+            relays.iter().filter(|r| !failed_relays.contains(r)).cloned().collect();
+
+        if remaining_relays.is_empty() {
+            return Err(anyhow!("No valid relays available"));
+        }
+
+        let selected_relay =
+            match remaining_relays.choose(&mut payjoin::bitcoin::key::rand::thread_rng()) {
+                Some(relay) => relay.clone(),
+                None => return Err(anyhow!("Failed to select from remaining relays")),
+            };
+
+        relay_state.lock().unwrap().set_selected_relay(selected_relay.clone());
+
+        let ohttp_keys =
+            payjoin::io::fetch_ohttp_keys(selected_relay.clone(), payjoin_directory.clone()).await;
+
+        match ohttp_keys {
+            Ok(_) => return Ok(selected_relay),
+            Err(error) => {
+                let status_code = error
+                    .to_string()
+                    .split_whitespace()
+                    .find_map(|code| http::StatusCode::from_str(code).ok())
+                    .expect("Could not find status code");
```

Comment:
I am doing some string splits here instead of changing the payjoin io `Error::UnexpectedStatusCode` It still feels a bit brittle, curious about other ideas you might have beyond making `InternalError` public.

Reply:
Making the error public is the way to go, perhaps with just an/some private variants

Reply:
Will revert this logic back to 6685f8528955ca7bb86b839db5e4494d8813a86b